File: java/L100/LeetCode0110.java
package L100;

import util.TreeNode;

public class LeetCode0110 {
    class Res {
        public int height;
        public boolean balance;
        public Res(int h,boolean b) {
            height = h;
            balance = b;
        }
    }
    public boolean isBalanced(TreeNode root) {
        return judge(root).balance;
    }
    private Res judge(TreeNode root) {
        if (root == null) {
            return new Res(0,true);
        }
        Res lr = judge(root.left);
        Res rr = judge(root.right);
        if (!lr.balance || !rr.balance || Math.abs(lr.height - rr.height) > 1) {
            return new Res(1+Math.max(lr.height, rr.height),false);
        } else {
            return new Res(1+Math.max(lr.height, rr.height),true);
        }
    }
}

File: cpp/L100/LeetCode0110.cc
#include "../heads/tree_node.h"
#include <algorithm>

class Solution {
    std::pair<bool, int> balanceTree(TreeNode *node) {
        if (node == nullptr) {
            return {true, 0};
        }
        auto [bl, nl] = balanceTree(node -> left);
        auto [br, nr] = balanceTree(node -> right);
        bool b = bl && br && std::abs(nl - nr) <= 1;
        return {b, std::max(nl, nr) + 1};
    }
public:
    bool isBalanced(TreeNode* root) {
        return balanceTree(root).first;
    }
};
